### PR TITLE
Prevent updating existing releases

### PR DESCRIPTION
**Pull request by baijum**
Prevent a partner from updating the released version of a Helm chart with the
same version number. Ref. #182

Jira: https://issues.redhat.com/browse/HELM-186

### CODE REVIEW

File: scripts/src/sanitycheckpr/sanitycheckpr.py
Comment (mmulholla) [inline scripts/src/sanitycheckpr/sanitycheckpr.py:41]: "elif match" should be "else" 
given match is True "pattern_match=True" is easier to follow.

File: scripts/src/sanitycheckpr/sanitycheckpr.py
Comment (mmulholla) [inline scripts/src/sanitycheckpr/sanitycheckpr.py:67]: tag_name is also determined in chartrepomanger.py, as a result it would be better to set tag_name as an output which chartrepomanger.py can use. This way we ensure both are using the same thing and we don't need to update two places if it ever changed for any reason. The same goes for entry_name.

File: scripts/src/sanitycheckpr/sanitycheckpr.py
Comment (mmulholla) [inline scripts/src/sanitycheckpr/sanitycheckpr.py:72]: To get here there is no entry for the chart in Index.yaml, so could we ever get a chart which has not entry but does have a release? If not I am not sure we need this check.

Comment (baijum) [reply scripts/src/sanitycheckpr/sanitycheckpr.py:72]: A tag cannot exist without a corresponding entry in the `index.yaml` in an ideal condition. Since tags can be created by other means, I thought it is better to safeguard against tag existence.

File: scripts/src/sanitycheckpr/sanitycheckpr.py
Comment (baijum) [reply scripts/src/sanitycheckpr/sanitycheckpr.py]: Done. bf4863c47c0f01ad2f4bfe80a388d432e6d621b9

File: scripts/src/sanitycheckpr/sanitycheckpr.py
Comment (baijum) [reply scripts/src/sanitycheckpr/sanitycheckpr.py:68]: Done. bf4863c47c0f01ad2f4bfe80a388d432e6d621b9

File: scripts/src/sanitycheckpr/sanitycheckpr.py
Comment (pedjak) [inline scripts/src/sanitycheckpr/sanitycheckpr.py:26]: the current and proposed code changes still make possible that a PR contains changes to multiple charts. Namely, the regex only looks if filepaths are starting with `charts/` prefix and have the defined depth.

IMHO, forbidding updates can be done a way simpler, even without checking index.yaml:

* Get list of files submitted in PR (that is already done in line 25)
* All file statuses should be set to `added` - otherwise, we can exit with error - if PR contains some files that are modified, it means that PR modifies some existing chat, and that chart is then already published, because it is a part of code repo.
* All files in the PR should have share the same common prefix

There is no need to download any `index.yaml` and the whole workflow logic becomes a way simpler.

Comment (pedjak) [reply scripts/src/sanitycheckpr/sanitycheckpr.py:26]: cc @mmulholla

Comment (baijum) [reply scripts/src/sanitycheckpr/sanitycheckpr.py:20]: Done. 5d5ac7e4ef53852d1935ce45c92b993980b7884d

Comment (mmulholla) [reply scripts/src/sanitycheckpr/sanitycheckpr.py:20]: It is possible to update a chart by adding a new file,  which this would now miss. Not sure if you can add a file which could affect the contents of the Index.yaml file but we would continue to create the release. The previous approach required more code to cover this issue.

Comment (mmulholla) [reply scripts/src/sanitycheckpr/sanitycheckpr.py:20]: maybe we just add a check that Chart.yaml is included in the PR?

Comment (baijum) [reply scripts/src/sanitycheckpr/sanitycheckpr.py:20]: Added checking the existence of chart and/or report as part of the PR:
0f8447bdd008cb36e49214fbf310db4e69feea04

File: docs/README.md
Comment (baijum) [reply docs/README.md:199]: See the status part of this PR. Currently, it's written, "All checks have failed". But as soon as we introduce one more check, the message could be either "Some checks were not successful" or "All checks have failed".
This is a GitHub UX.

File: scripts/src/sanitycheckpr/sanitycheckpr.py
Comment (mmulholla) [reply scripts/src/sanitycheckpr/sanitycheckpr.py:20]: I don't think  the logic is quite right yet.

report found, need to ensure no other files are added unless chart.yaml is present.  
report found, tarball found, need to ensure no other files are added.
no report, must ensure either chart.yaml is included, or ,tgz is the only file.

Comment (baijum) [reply scripts/src/sanitycheckpr/sanitycheckpr.py:20]: These combination checks are present as part of `Verify PR - generate report`  step.
In the sanity check, we need to ensure only addition is happening. And you pointed what if the user adds an extra file to the existing chart. To last commit ensure "Chart.yaml" is in the add "added" list (Same with report and chart).  This was we can ensure only addition is part of the PR. And if it's all addition there is an addition to one of the above three files.

Comment (mmulholla) [reply scripts/src/sanitycheckpr/sanitycheckpr.py:20]: I thought the idea here was to prevent updating an existing chart by detection in sanity check.
1. I can publish a chart without a report and then add a report later, this should be caught but is not in the current logic. 
2. I can publish a chart using just a report, then add chart source or a tarball, this should be caught but is not in the current logic.   
3. Not sure what it means if I publish a chart using source and then try adding a tarball, but should probably also be caught in this logic.

Beginning to think this concept of making sure all contents of a PR are additions is flawed, ad the old logic was better.

Comment (baijum) [reply scripts/src/sanitycheckpr/sanitycheckpr.py:20]: @mmulholla I agree that this logic is flawed! I am with going back to the old approach in PR.
@pedjak Any comments?

Comment (pedjak) [reply scripts/src/sanitycheckpr/sanitycheckpr.py:20]: good catch @mmulholla 

In order to fix the mentioned issues, we should add one more check:

* the chart directory must not exist in the target branch

Comment (baijum) [reply scripts/src/sanitycheckpr/sanitycheckpr.py:20]: @mmulholla @pedjak @dperaza4dustbit This discussion is going back and forth multiple times. I need a confirmation whether I should revert my changes as suggested by @mmulholla or try the last suggestion given by @pedjak .

Comment (mmulholla) [reply scripts/src/sanitycheckpr/sanitycheckpr.py:20]: Personally I prefer checking Index.yaml content because we are checking exactly what we are trying to avoid, rather than checking some upstream status only to find some corner case that it does not cover.  Checking the chart version directory does not already exist certainly makes for simpler code but I would worry that we could merge a chart and then fail in the release process for some reason. In this case one solution we would now prevent is  to trigger the release process again using a new PR with a irrelevant change.  I am assuming that chart content which passes build and verify could never cause a release proces failure.

Comment (mmulholla) [reply scripts/src/sanitycheckpr/sanitycheckpr.py:20]: @predrag please respond to my comment. Will leave the final decision to you.

Comment (pedjak) [reply scripts/src/sanitycheckpr/sanitycheckpr.py:20]: > please respond to my comment. Will leave the final decision to you.

I am fine if you wanna take this route.

Comment (dperaza4dustbit) [reply scripts/src/sanitycheckpr/sanitycheckpr.py:20]: Why don't we define a set of test cases we want to pass to call this PR done. Then we check against that and move pending things to a different story.

**Comment by mmulholla** [comment]
@baiju In the new code build.yml exits early if sanity check fails but then prepare_pr_comment.py looks for messages from sanity check. I think previously we eliminated that early exit and checked more often it passed.

**Comment by baijum** [comment]
> @baijum In the new code build.yml exits early if sanity check fails but then prepare_pr_comment.py looks for messages from sanity check. I think previously we eliminated that early exit and checked more often it passed.

Before every exit from the sanity check, an error message has been set with a print statement like this:
`::set-output name=sanity-error-message::...`

The only possible scenario where the message not set is when installing dependencies fails.
For the last few hours, it's happening as there is an ongoing infra failure: https://status.python.org/

File: .github/workflows/build.yml
Comment (mmulholla) [inline .github/workflows/build.yml:147]: We set a variable here for a sanity check error but if there is a sanity check error we would never get here because of the  "Reflect on Sanity Check PR Content" step which exits on sanity check failure well before we get here.

Comment (baijum) [reply .github/workflows/build.yml:147]: There is an `if: always()` condition (line. 145), so this is going to run irrespective of any other previous job fails or not.

Comment (mmulholla) [reply .github/workflows/build.yml:147]: So "exit 1" doesn't exit, perfectly obvious! Got it.

**Review by dperaza4dustbit — APPROVED** [review]

**Comment by dperaza4dustbit** [comment]
Before we merge this, lets make sure there is not regression with an end to end submission test.

**Review by mmulholla — APPROVED** [review]

**Pull request merged**
